### Recursion, the Fibonacci Sequence and Memoization || Python Tutorial || Learn Python Programming

https://www.youtube.com/watch?v=Qk0zUZW-U_M

In [16]:
from time import process_time

In [17]:
# let's first define a regular fibonacci function using recursion
def fib(n):
    if n<=2: return 1
    return fib(n-1) + fib(n-2)

In [23]:
start_time = process_time()
for i in range(1,41):
    print(i,':',fib(i))
print('Time to process',process_time()-start_time)

1 : 1
2 : 1
3 : 2
4 : 3
5 : 5
6 : 8
7 : 13
8 : 21
9 : 34
10 : 55
11 : 89
12 : 144
13 : 233
14 : 377
15 : 610
16 : 987
17 : 1597
18 : 2584
19 : 4181
20 : 6765
21 : 10946
22 : 17711
23 : 28657
24 : 46368
25 : 75025
26 : 121393
27 : 196418
28 : 317811
29 : 514229
30 : 832040
31 : 1346269
32 : 2178309
33 : 3524578
34 : 5702887
35 : 9227465
36 : 14930352
37 : 24157817
38 : 39088169
39 : 63245986
40 : 102334155
Time to process 48.494179


<font color='yellow'>It will take forever if we try to compute the same for 'i' in range(1,100)!</font>

In [19]:
fibonacci_cash = {}

In [20]:
def fibonacci(n):
    if n in fibonacci_cash:
        return fibonacci_cash[n]
    if n <= 2:
        value = 1
    else:
        value = fibonacci(n-1) + fibonacci(n-2)
    fibonacci_cash[n] = value
    return value
        

In [24]:
start_time=time.process_time()
for i in range(1,41):
    print(i,':',fibonacci(i))
print('Time to process:',process_time()-start_time)

1 : 1
2 : 1
3 : 2
4 : 3
5 : 5
6 : 8
7 : 13
8 : 21
9 : 34
10 : 55
11 : 89
12 : 144
13 : 233
14 : 377
15 : 610
16 : 987
17 : 1597
18 : 2584
19 : 4181
20 : 6765
21 : 10946
22 : 17711
23 : 28657
24 : 46368
25 : 75025
26 : 121393
27 : 196418
28 : 317811
29 : 514229
30 : 832040
31 : 1346269
32 : 2178309
33 : 3524578
34 : 5702887
35 : 9227465
36 : 14930352
37 : 24157817
38 : 39088169
39 : 63245986
40 : 102334155
Time to process: 0.005557000000010248


### As you can see the difference in time to process is HUGE!!

* Time to process <font color='yellow'>48.494179<font>
* Time to process: <font color='yellow'>0.005557000000010248<font>


In [29]:
#Let's simplyfy the original code by adding a caching function
from functools import lru_cache

@lru_cache(maxsize=1000)
def fibo(n):
    if n<=2: return 1
    return fibo(n-1) + fibo(n-2)

In [30]:
start_time=time.process_time()
for i in range(1,41):
    print(i,':',fibo(i))
print('Time to process:',process_time()-start_time)

1 : 1
2 : 1
3 : 2
4 : 3
5 : 5
6 : 8
7 : 13
8 : 21
9 : 34
10 : 55
11 : 89
12 : 144
13 : 233
14 : 377
15 : 610
16 : 987
17 : 1597
18 : 2584
19 : 4181
20 : 6765
21 : 10946
22 : 17711
23 : 28657
24 : 46368
25 : 75025
26 : 121393
27 : 196418
28 : 317811
29 : 514229
30 : 832040
31 : 1346269
32 : 2178309
33 : 3524578
34 : 5702887
35 : 9227465
36 : 14930352
37 : 24157817
38 : 39088169
39 : 63245986
40 : 102334155
Time to process: 0.0045709999999985484


### This is much much better, and easier!